FV3/coarse_graining/coarse_graining.F90

```diff
@@ -300,6 +309,55 @@ subroutine weighted_block_edge_average_y_3d_field_2d_weights(weights, fine, coar
     enddo
   end subroutine weighted_block_edge_average_y_3d_field_2d_weights
 
+  subroutine vertically_remap_field(phalf_in, field, phalf_out, ptop, field_out)
+    real, intent(in) :: phalf_in(is:ie,js:je,1:npz+1), phalf_out(is:ie,js:je,1:npz+1)
+    real, intent(in) :: field(is:ie,js:je,1:npz)
+    real, intent(in) :: ptop
+    real, intent(out) :: field_out(is:ie,js:je,1:npz)
+
+    integer :: kn, km, kord, iv, j, q2
+
+    kn = npz
+    km = npz
+
+    ! Hard code values of kord and iv for now
+    kord = 1
+    iv = 1
+    q2 = 1
+
+    do j = js, je
+       call mappm(km, phalf_in(is:ie,j,:), field(is:ie,j,:), kn, &
+            phalf_out(is:ie,j,:), field_out(is:ie,j,:), is, ie, iv, kord, ptop)
+    enddo
+  end subroutine vertically_remap_field
+
+  subroutine block_upsample_2d(coarse, fine)
+    real, intent(in) :: coarse(is_coarse:ie_coarse,js_coarse:je_coarse)
+    real, intent(out) :: fine(is:ie,js:je)
+
+    integer :: i, j, i_coarse, j_coarse, offset
+
+    offset = coarsening_factor - 1
+    do i = is, ie, coarsening_factor
+      i_coarse = (i - 1) / coarsening_factor + 1
+      do j = js, je, coarsening_factor
+          j_coarse = (j - 1) / coarsening_factor + 1
+          fine(i:i+offset,j:j+offset) = coarse(i_coarse, j_coarse)
+      enddo
+    enddo
+  end subroutine block_upsample_2d
+
+  subroutine block_upsample_3d(coarse, fine)
+    real, intent(in) :: coarse(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz+1)
+    real, intent(out) :: fine(is:ie,js:je,1:npz+1)
```

Comment:
Is it `npz+1` because this is only intended upsample variables on level interfaces, i.e. "phalf"? Maybe should say that in subroutine name, or take # of vertical levels as argument.

Reply:
Good thought -- I agree this would be cleaner with a vertical levels argument.  Updated.